### PR TITLE
Complete restructure - mother of all PRs

### PR DESCRIPTION
**Pull request by benc-uk**
Complete restructure and refactor. this is a monster PR
- The command structure now implements [the proposal in in the wiki](https://github.com/aztfmod/rovergo/wiki/Rover-v2-Draft-Proposal#june-10th---proposal-only). This means that action verbs are top level commands
- the `ci`, `cd`, `landingzone` and `launchpad` commands are gone
- Custom commands now replace `ci` and `ci_tasks` directory is replaced with `custom_actions`
- Custom actions can be any external executable, and they are dynamically loaded and extend the rover CLI a bit like plugins, e.g.  See the custom_actions folder for two examples `rover linter` and `rover finder`
- The operational mode is determined by the presence of either the `--config-dir` or the `--config-file` flags, which changes between adhoc or single level mode (without a symphony YAML) and multi-level mode (with a symphony YAML)
- All actions are now implemented as via an interface `Action` with `TerraformAction` and `CustomAction` being implemenatons, and each of the terraform actions are separate implementations e.g. `PlanAction` and `ApplyAction`.
- Actions must implement the `Execute(o *Options)` which is how they are run.

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -45,8 +55,75 @@ func GetVersion() string {
...
+				configFile, _ := cmd.Flags().GetString("config-file")
+				configPath, _ := cmd.Flags().GetString("config-dir")
+
+				// Handle the user trying to use both configPath and configFile or neither!
+				if configPath == "" && configFile == "" {
+					_ = cmd.Help()
+					os.Exit(0)
+				}
+				if configPath != "" && configFile != "" {
+					cobra.CheckErr("--config and --config-dir options must not be combined, specify only one")
```
Comment (sebastus) [inline cmd/root.go:92]: --config-file and --config-dir  ??

File: pkg/landingzone/cobra.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	cobra.CheckErr(err)
+
+	// Normally cobra would provide automatic defaults but we are using it in a weird way
+	if level == "" {
+		cobra.CheckErr("--level must be supplied when not using a config file")
+	}
+	if sourcePath == "" {
+		cobra.CheckErr("--source option must be supplied when not using a config file")
+	}
+	if ws == "" {
```
Comment (sebastus) [inline pkg/landingzone/cobra.go:39]: extract defaults to const?

File: pkg/landingzone/cobra.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+		ws = "tfstate"
+	}
+	if env == "" {
+		env = "sandpit"
```
Comment (sebastus) [inline pkg/landingzone/cobra.go:43]: same

File: pkg/landingzone/landingzone.go [hunk 4]
```diff
@@ -29,15 +28,13 @@ const secretTenantID = "tenant-id"
 const secretLowerSAName = "lower-storage-account-name"
```
Comment (sebastus) [inline pkg/landingzone/landingzone.go:28]: export consts? - they might be needed elsewhere - for example I used secretTenantID in a test

**Review by sebastus — CHANGES_REQUESTED** [review]
so much better structurally! great work!

**Comment by sebastus** [comment]
Can we further isolate config and execute? Something like this:

pkg/landingzone/cobra.go:
`func RunFromCLI(cmd *cobra.Command, action Action) *[]Options {
...
   return []Options{opt}
}`

pkg/symphony/cobra.go
`func RunFromConfig(cmd *cobra.Command, action landingzone.Action) *[]Options {
...
	if levelName == "" {
		return conf.runAll(action, dryRun)  <-- returns array of Options
	}
...
	return conf.RunLevel(*level, action, dryRun)  <-- returns array of Options
`

root.go
`var options []Options
if configFile != "" {
	options = symphony.RunFromConfig(cmd)
} else {
	options = landingzone.RunFromCLI(cmd)
}

for _, optionSet := range options {
	_ = action.Execute(optionSet)
}
console.Success("Rover has finished")
os.Exit(0)
`

Then, a test can do the initial bit to build the options array and assert on its contents without running execute. An integration test can go on to run Execute and check the actual resources.

File: pkg/landingzone/cobra.go (hunk 3 above)
Comment (benc-uk) [reply pkg/landingzone/cobra.go:43]: Good idea

File: cmd/root.go (hunk 1 above)
Comment (benc-uk) [reply cmd/root.go:92]: Good catch

File: pkg/landingzone/landingzone.go (hunk 4 above)
Comment (benc-uk) [reply pkg/landingzone/landingzone.go:28]: done

**Review by sebastus — APPROVED** [review]
lgtm

let's merge it and deal with whatever's left in another PR.

**Review by sebastus — APPROVED** [review]
lgtm

**Pull request merged**